awsshell/wizard.py

```diff
@@ -1,53 +1,105 @@
+        services = self._loader.list_available_services(type_name=name)
+        model = self._loader.load_service_model(services[0], name)
+        return Wizard(model)
+
+
+class Wizard(object):
+    """The main wizard object containing all of the stages, the environment,
+    botocore sessions, and the logic to drive the wizards.
+    """
+
+    def __init__(self, spec=None):
+        """Constructs a new Wizard
+
+        :type spec: dict
+        :param spec: (Optional) Constructs the wizard by loading the given
+        model specification.
+        """
+        self.env = Environment()
+        self._session = botocore.session.get_session()
+        self._cached_creator = index.CachedClientCreator(self._session)
+        if spec:
+            self.load_from_dict(spec)
 
-    # Loads the wizards from the spec in dict form
     def load_from_dict(self, spec):
+        """Loads the model specification from a dict, populating the
+        wizard object.
+
+        :type spec: dict
+        :param spec: The model specification.
+
+        :raises: :class:`WizardException`
+        """
         self.start_stage = spec.get('StartStage', None)
         if not self.start_stage:
             raise WizardException("Start stage not specified")
         self.stages = {}
         for s in spec['Stages']:
-            stage = Stage(s, self._env)
+            stage = Stage(s, self)
             self.stages[stage.name] = stage
 
-    # TODO temporary loader from file, will likely deprecate this when
-    # there's a systemic way of loading all wizards in specified locations
-    def load_from_json(self, filename):
-        with open(filename, 'r') as f:
-            spec = json.load(f)
-            self.load_from_dict(spec)
-
-    # Performs the basic loop for progressing stages
     def execute(self):
+        """Runs the wizard. Executes Stages until a final stage is reached.
+
+        :raises: :class:WizardException
+        """
         current_stage = self.start_stage
         while current_stage:
             stage = self.stages.get(current_stage, None)
             if not stage:
                 raise WizardException("Stage not found: %s" % current_stage)
             stage.execute()
             current_stage = stage.get_next_stage()
-        print(self._env)
+        # TODO decouple wizard from all I/O
+        sys.stdout.write(str(self.env)+'\n')
 
 
 class Stage(object):
+    """The Stage object contains the meta information for a stage and logic
+    required to perform all steps present.
+    """
+
+    def __init__(self, spec, wizard):
+        """Constructs a new Stage object.
 
-    def __init__(self, spec, env):
-        self._wizard_env = env
+        :type spec: dict
+        :param spec: The stage specification model.
+
+        :type wizard: :class:`Wizard`
+        :param wizard: The wizard that this stage is a part of.
+        """
+        self._wizard = wizard
         self.name = spec.get('Name', None)
```

Comment:
I think I would rather see these as explicit constructor arguments. You can use your existing wizard factory to populate those.